include/revng/PTML/CommentPlacementHelper.h

```diff
@@ -0,0 +1,339 @@
+  // safe) to copy
+  std::unordered_map<StatementType, CommentList> ResultMap;
+  CommentList HomelessComments;
+
+public:
+  CommentPlacementHelper() = default;
+  CommentPlacementHelper(const model::Function &Function, auto &&Graph) {
+    if (Function.Comments().empty()) {
+      // No comments in this function, nothing to map
+      return;
+    }
+
+    // Compute a domination tree for the given graph to reduce duplicate
+    // comments: if a suitable statement is already dominated by another one,
+    // we can skip that emission.
+    using InternalNodeType = std::remove_pointer_t<NodeType>;
+    using DominatorTreeType = llvm::DominatorTreeBase<InternalNodeType, false>;
+    using DTNode = const llvm::DomTreeNodeBase<InternalNodeType> *;
+    DominatorTreeType DominatorTree;
+    DominatorTree.recalculate(Graph);
+
+    std::vector<std::pair<Score, StatementLocationType>> Scores;
+    Scores.resize(Function.Comments().size());
+
+    // First build a score map - select the best location for each comment
+    for (auto Node : llvm::depth_first(DominatorTree.getRootNode())) {
+      for (const auto &Statement : Trait::getStatements(Node->getBlock())) {
+        StatementLocationType Location = Trait::getAddresses(Statement);
+
+        for (auto &&[Index, Comment] : llvm::enumerate(Function.Comments())) {
+
+          // Use the Tversky Index as the scoring function.
+          // Note the parameters controlling the weights of the sets: the higher
+          // a value is, the bigger the relative negative impact of
+          // the corresponding set having _extra_ elements compared to the other
+          // one.
+          //
+          // We want this because a line having unaccounted for addresses is a
+          // way bigger red flag that a (potentially user edited) comment having
+          // them.
+          auto Score = tverskyIndex<1, 2>(Comment.Location(), Location);
+          if (Score.has_value())
+            if (shouldReplace(Scores[Index], { *Score, Location }))
+              Scores[Index] = { *Score, Location };
+        }
+      }
+    }
+
+    // Process homeless comments so they don't interfere with the others
+    for (auto &&[I, Comment] : llvm::enumerate(Function.Comments()))
+      if (Scores[I].first.Numerator == 0)
+        HomelessComments.emplace_back(I, false, Score{}, &Comment.Location());
+
+    // Use the tree above to build a node-to-comment-list map.
+    struct TreeVisitor {
+      std::unordered_map<StatementType, CommentList> &ResultMap;
+      const model::Function &Function;
+      const std::vector<std::pair<Score, StatementLocationType>> &Scores;
+
+      void operator()(const DTNode &Node) {
+        llvm::SmallVector<bool> ShouldSkip;
+        ShouldSkip.resize(Scores.size());
+
+        return impl(Node, std::move(ShouldSkip));
+      }
+
+    private:
+      void impl(const DTNode &Node, llvm::SmallVector<bool> ShouldSkip) {
```

Comment:
I'd rename `ShouldSkip` to `AssignedInThisBranch`

Reply:
Done.